---
title: "Model building fecal metagenomics dataset"
author: "Ece Kartal"
date: "5.12.2023"
output:
  pdf_document: default
  html_document: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here, we will use SIAMCAT for logistic regression model building. 
This is only a placeholder since metadata is not yet finalised. 
Based on the final set and numbers, the groups compared have to be updated 
via **case** variable in siamcat function below. 
SIAMCAT needs a feature matrix (matrix or data.frame) 
features (in rows) samples (in columns)
metadata in a data.frame, samples as row names

```{r, include=FALSE}
# load library
library("tidyverse")
library("matrixStats")
library("SIAMCAT")
library("ggrepel")
library("dplyr")
library("plyr")
library("readxl")
# check siamcat version
packageVersion("SIAMCAT")
PARAM <- list()
PARAM$folder.R <- paste0(getwd(), "/")
PARAM$folder <- gsub("src/", "", PARAM$folder.R)
PARAM$folder.input <- paste0(PARAM$folder, "input/")
PARAM$folder.Rdata <- paste0(PARAM$folder, "Rdata/")
PARAM$folder.data <- paste0(PARAM$folder, "data/")
PARAM$folder.output <- paste0(PARAM$folder, "output/metaG/")

# load data metag motu and genes
# load metadata 
load(paste0(PARAM$folder.Rdata,"metaBatch123.Rdata"))
# load motu files
load(paste0(PARAM$folder.Rdata,"metag.motus.batch123.Rdata"))

```

```{r}
featTable=motu.rel
metaTable=meta
# choose the meta variables to test for confounding
metatest = c("CHEMO", "IMMUN", "TARGET",  "Age", "Sex","Vital_status", "Timepoint",   
             "days_to_death" , "ReadCount (Homo)", "lib_size_factor", "lib_Homo_factor", 
             "Cause of death", "age_factor")
```

###############################################################################
### Overview
Function to run siamcat with centered log transformation (clr) normalization
10 fold cross validation and 10 fold re-sampling
Confounder check based on fisher exact test
###############################################################################

```{r}
# norm:"rank.unit", "rank.std","log.std", "log.unit", "log.clr", "std", "pass"
# ml: 'lasso', 'enet', 'ridge', 'lasso_ll', 'ridge_ll', 'randomForest'
# featTable: relative abundance table
# metaTable:metadata
# label:column name of the comparison 
# fileName: filenema when saving all resutls 
# case: comparison e.g cancer vs control, responder vs nonresponder. 
# I am not using responder since we dont have the proper metadata yet. 

runsiamcat <- function(featTable, metaTable, label, fileName, case, ml, norm){
  dim(featTable)
  # create SIAMCAT object and classify
  siamcat <- siamcat(feat=featTable, meta=metaTable, label=label, case=case)
  
  # filter based on abundance 
  siamcat <- filter.features(siamcat, filter.method = 'abundance', 
                            cutoff=0.001, verbose=3)
  check.confounders(siamcat, fn.plot = paste0(PARAM$folder.output, 
                            fileName, '.confounders.pdf'),
                            meta.in=metatest, verbose = 3)
        
  # normalize with log.clr
  siamcat <- normalize.features(siamcat, norm.method = norm, 
                                feature.type = 'filtered',
                                norm.param = list(log.n0=1e-05, sd.min.q=1))
                                      
  # compute associations 
  siamcat <- check.associations(siamcat, feature.type = 'normalized')
        
  # train model
  siamcat <- create.data.split(siamcat, num.folds = 5, num.resample = 5) 
  # has to be 10 and 10, I only use 5 for testing purpuses 
  siamcat <- train.model(siamcat, method = ml, verbose = 2)
  siamcat <- make.predictions(siamcat)
  siamcat <- evaluate.predictions(siamcat)    
  print(siamcat@eval_data$auroc)
  # evaluation plot
  model.evaluation.plot(siamcat, fn.plot = paste0(PARAM$folder.output, Sys.Date(), '.',
                                                  fileName, '.eval.plot.pdf'))
  model.evaluation.plot(siamcat)
  # interpretation plot
  model.interpretation.plot(siamcat, fn.plot = paste0(PARAM$folder.output, 
                           Sys.Date(), '.', fileName,'.interpret.plot.pdf'),
                           consens.thres = 0.5,
                           heatmap.type = 'zscore')
  # interpretation plot to print
  model.interpretation.plot(siamcat)
  # save siamcat object
  save(siamcat, file = paste0(PARAM$folder.output, fileName, '.siamcat.Rdata'))
  return(siamcat) 
}

```

###############################################################################
### MetaG Taxonomic Modelling
###############################################################################

```{r}
# prepare and subset proper meta and feat
# I subset only T1/T2 for some meaningful comparisons
metas=as.data.frame(meta) %>%
 filter(Timepoint %in% "T1/T2")
overlap_columns <- intersect(metas$ID, colnames(featTable))
metas_subset <- metas %>%
  filter(ID %in% overlap_columns)
metas_subset <- metas_subset[!duplicated(metas_subset$ID), ]
rownames(metas_subset) <- metas_subset$ID

# subset featTable
featTable <- featTable[, colnames(featTable) %in% overlap_columns]

# Check if row names are identical to column names in 'feat'
row_names_identical <- all(rownames(metas_subset) %in% 
                             colnames(featTable)) && length(rownames(metas_subset)) == ncol(featTable)
# Print the result
print(row_names_identical)

```

```{r,message=FALSE, warning=FALSE, echo=FALSE}
# run siamcat fpor example metavariables
# normalisation and logistic regression method has to be selected based on the data and final question, this is only as an example. 

runsiamcat(featTable, metas_subset, "Sex", "Sex", "Female", 'ridge', "log.std" ) 
runsiamcat(featTable, metas_subset, "Vital_status", "Vital_status", "0", 'ridge', "log.std" ) 
```

###############################################################################
### Add confounders to metaG models
Here we can test if specific clinical variable have additional informative power
to microbiome data. It has to be tested with full metadata
###############################################################################

```{r, message=FALSE, warning=FALSE, echo=FALSE}
# load models
load(paste0(PARAM$folder.output, "Sex.siamcat.Rdata"))
siamcat.sex <- siamcat

# add confounders to model

add.meta <- function(x, n) {
  x <- add.meta.pred(x, pred.names = n, verbose = 3)
  x <- train.model(x, method = 'ridge', verbose = 2, perform.fs = TRUE)
  x <- make.predictions(x)
  x <- evaluate.predictions(x)
  return(x)
}

# combine with naive model
siamcat.vit <- add.meta(siamcat.sex, 'Vital_status')

model.evaluation.plot('Only Sex model'= siamcat.sex,
                      'Vital status included model'= siamcat.vit)

model.interpretation.plot(siamcat.vit)

# save pdf
model.evaluation.plot('Only Sex model'= siamcat.sex,
                      'Vital status included model'= siamcat.vit,
                      fn.plot = paste0(PARAM$folder.output,  Sys.Date(),
                                       '.confounders.interpret.pdf'))

# save confounder Rdata
save(siamcat.sex, siamcat.vit,
     file = paste0(PARAM$folder.files, Sys.Date(), 'confounder.RData'))
```

Here I only showcase 2 examples, with final metadata and dataset this analysis has to be repeated btw responders and non-responders.
```{r session_info}
sessionInfo()
```


